### PR TITLE
Friends and enemies overhaul

### LINKED ISSUE #624: Detective uses "friends"
The detective picks someone and finds out if they are friends

This will use the same mechanic as the philosopher

Jester, Hater, and Doomsayer will appear innocent because they can win with the detective.

This is better for the detective, because it means if someone is sus, you need them killed to win

### PR DESCRIPTION
**Pull request by ItsSammyM**
closes #624

### CODE REVIEW

File: server/src/game/end_game_condition.rs [hunk 1]
```diff
@@ -2,26 +2,113 @@ use super::{role_list::Faction, role::Role};
...
+            EndGameCondition::Mafia,
+            EndGameCondition::Cult,
+            EndGameCondition::Arsonist,
+            EndGameCondition::Werewolf,
+            EndGameCondition::Ojo,
+            EndGameCondition::Death,
+            EndGameCondition::Politician
+        ]
+    }
+    ///either return Some(EndGameCondition) or None (if the game is not over yet)
```
Comment (Jack-Papel) [inline server/src/game/end_game_condition.rs:31]: Why?

File: server/src/game/end_game_condition.rs [hunk 2]
```diff
@@ -2,26 +2,113 @@ use super::{role_list::Faction, role::Role};
...
+        //if nobody is left to hold game hostage
+        if !living_roles.iter().any(|role|EndGameCondition::keeps_game_running(*role)){
+            return Some(EndGameCondition::Draw);
+        }
+
+        //find one end game condition that everyone agrees on
+        for end_game_condition in EndGameCondition::all() {
+            let mut all_agree = true;
+            for living_role in &living_roles {
+                if !EndGameCondition::keeps_game_running(*living_role) {continue;}
```
Comment (Jack-Papel) [inline server/src/game/end_game_condition.rs:48]: You already filtered all of these on like 40. This line is not needed

Comment (Jack-Papel) [reply server/src/game/end_game_condition.rs:48]: Nvm, I am dumb

File: server/src/game/end_game_condition.rs [hunk 3]
```diff
@@ -2,26 +2,113 @@ use super::{role_list::Faction, role::Role};
...
+    Arsonist,
+    Werewolf,
+    Ojo,
+
+    Death,
+    Politician,
+
+    Draw
 }
 impl EndGameCondition {
```
Comment (Jack-Papel) [inline server/src/game/end_game_condition.rs:18]: It really seems like now, this represents a "win condition", am I correct?

File: server/src/game/end_game_condition.rs [hunk 4]
```diff
@@ -2,26 +2,113 @@ use super::{role_list::Faction, role::Role};
...
+
+        //if nobody is left to hold game hostage
+        if !living_roles.iter().any(|role|EndGameCondition::keeps_game_running(*role)){
+            return Some(EndGameCondition::Draw);
+        }
+
+        //find one end game condition that everyone agrees on
+        for end_game_condition in EndGameCondition::all() {
+            let mut all_agree = true;
+            for living_role in &living_roles {
```
Comment (Jack-Papel) [inline server/src/game/end_game_condition.rs:47]: I think you should replace this for loop with a `living_roles.iter()...` thing. Personally the way it's written seems a little obtuse -- or at least it was hard for me to understand at first

File: server/src/game/mod.rs [hunk 5]
```diff
@@ -270,24 +270,17 @@ impl Game {
...
-                .collect::<std::collections::HashSet<EndGameCondition>>().into_iter().collect::<Vec<EndGameCondition>>();
-
-        //if there are no teams left and multiple wildCards alive then the game is not over
-        if
-            remaining_teams.is_empty() && 
-            PlayerReference::all_players(self)
-                .filter(|p|p.alive(self) && p.role_state(self).role() == role::Role::Wildcard)
-                .count() > 1 
+                .filter_map(|p|
+                    if p.alive(self){Some(p.role(self))}else{None})
```
Comment (Jack-Papel) [inline server/src/game/mod.rs:276]: This whitespace is so evil, atp I'd rather have a separate filter and map if you're gonna write your if statements like this. We're not in the interest of saving bytes here

File: server/src/game/role/common_role.rs [hunk 6]
```diff
@@ -134,21 +134,16 @@ pub(super) fn get_current_receive_chat_groups(game: &Game, actor_ref: PlayerRefe
     out
 }
 
+///Only works for roles connected to a faction/team
 pub(super) fn get_won_game(game: &Game, actor_ref: PlayerReference) -> bool {
```
Comment (Jack-Papel) [inline server/src/game/role/common_role.rs:138]: `get_team_won_game`

File: server/src/game/role/common_role.rs [hunk 7]
```diff
@@ -134,21 +134,16 @@ pub(super) fn get_current_receive_chat_groups(game: &Game, actor_ref: PlayerRefe
...
-        .collect::<Vec<_>>();
-
-    //if the only remaining endgame conditions are none and or the actor_ref's role's endgamecondition then true
-    remaining_end_game_conditions.iter().all(|end_game_condition|
-        *end_game_condition == EndGameCondition::None || 
-        *end_game_condition == actor_ref.role(game).end_game_condition()
-    )
+    let living_roles = PlayerReference::all_players(game)
+        .filter_map(|p| 
+            if p.alive(game) {Some(p.role(game))}else{None}
```
Comment (Jack-Papel) [inline server/src/game/role/common_role.rs:141]: See above whitespace comment

File: server/src/game/role/common_role.rs [hunk 8]
```diff
@@ -134,21 +134,16 @@ pub(super) fn get_current_receive_chat_groups(game: &Game, actor_ref: PlayerRefe
...
+        ).collect();
+    let end_game_condition = EndGameCondition::game_is_over(living_roles);
+    if let Some(end_game_condition) = end_game_condition {
```
Comment (Jack-Papel) [inline server/src/game/role/common_role.rs:144]: nitpick but you can inline this and you should

File: server/src/game/role/common_role.rs (hunk 6 above)
Comment (Jack-Papel) [reply server/src/game/role/common_role.rs:138]: Also, why does this only work for roles connected to a faction/team? Wouldn't this work just fine for arsonist?

File: server/src/game/role/mod.rs [hunk 9]
```diff
@@ -309,13 +305,9 @@ impl Role{
     }
     pub fn has_suspicious_aura(&self, _game: &Game)->bool{
         match self {
-            Role::Politician => true,
             _ => false,
         }
     }
```
Comment (Jack-Papel) [inline server/src/game/role/mod.rs:310]: Unnecessary match. Is your linter not yelling at you for this?

File: server/tests/role.rs [hunk 10]
```diff
@@ -100,18 +100,49 @@ fn sheriff_basic() {
     kit::scenario!(game in Night 1 where
```
Comment (Jack-Papel) [inline server/tests/role.rs:100]: This test is called sheriff_basic. This is not basic at all. Move the minion, jester, and politician tests to another test case.

File: server/tests/role.rs [hunk 11]
```diff
@@ -134,11 +165,12 @@ fn seer_basic() {
         mafia2: Informant,
```
Comment (Jack-Papel) [inline server/tests/role.rs:165]: Again, no longer basic

**Review by Jack-Papel — CHANGES_REQUESTED** [review]
Get alex to review your wiki stuff. I didn't read to see if it was right. Also, I didn't test this

File: server/src/game/end_game_condition.rs (hunk 1 above)
Comment (ItsSammyM) [reply server/src/game/end_game_condition.rs:31]: If you are responding to the comment
I will use this later to send a message saying who won and stuff

File: server/src/game/end_game_condition.rs (hunk 3 above)
Comment (ItsSammyM) [reply server/src/game/end_game_condition.rs:18]: idk, Draw isnt a win condition. Its the state of the game when it ends

File: server/src/game/end_game_condition.rs (hunk 4 above)
Comment (ItsSammyM) [reply server/src/game/end_game_condition.rs:47]: I agree, fix

File: server/src/game/role/common_role.rs (hunk 6 above)
Comment (ItsSammyM) [reply server/src/game/role/common_role.rs:138]: arsonist is a team. I need help naming this stuff

File: server/src/game/role/mod.rs (hunk 9 above)
Comment (ItsSammyM) [reply server/src/game/role/mod.rs:310]: what about when i add the four billion future roles that have a suspicious aura by default

Comment (Jack-Papel) [reply server/src/game/role/mod.rs:310]: Bro never heard of `#[allow(blablabla)]`

File: client/src/resources/lang/en_us.json [hunk 12]
```diff
@@ -779,7 +779,7 @@
...
     "wiki.article.standard.friends.title": "Friends",
-    "wiki.article.standard.friends.text": "If two players can win the same game, they are friends.\n- A godfather can not win in the same game as a detective, they both need the other to lose, so they are enemies.\n - A jester can win in the same game as a detective, so they are friends.\n\n The philosopher and detective can use their abilities to see if people are friends or enemies.",
+    "wiki.article.standard.friends.text": "If two players can win the same game, they are friends.\n - A jester can win in the same game as a detective, so they are friends\n- An Apostle can win in the same game as a disciple so they are friends\n- A godfather can't win in the same game as a detective, they both need the other to lose, so they are enemies. However, the godfather has an innocent aura, so they might appear as friends\n The philosopher and detective can use their abilities to see if people are friends or enemies.",
```
Comment (Jack-Papel) [inline client/src/resources/lang/en_us.json:782]: This still mentions godfather, which has an innocent aura

File: client/src/resources/lang/en_us.json [hunk 13]
```diff
@@ -779,9 +779,9 @@
...
+    "wiki.article.standard.friends.text": "If two players can win the same game, they are friends.\n- A godfather can not win in the same game as a detective, they both need the other to lose, so they are enemies.\n - A jester can win in the same game as a detective, so they are friends.\n\n The philosopher and detective can use their abilities to see if people are friends or enemies.",
     "wiki.article.standard.enemies.title": "Enemies",
-    "wiki.article.standard.enemies.text": "A Philosopher's result which indicates that the two players they visited can not win together.",
+    "wiki.article.standard.enemies.text": "If two players need the other to lose in order for themselves to win, they are enemies.\n- A Politician can't win alongside a mayor, so they are enemies.\n- A godfather can't win in the same game as a jailor, so they are enemies.\n- A jester can win in the same game as a detective, so they are friends.\n- An Apostle can't win with a godfather, so they are enemies.\n\n The philosopher and detective can use their abilities to see if people are friends or enemies.",
```
[the remark this answers is not in the document]
Comment (Jack-Papel) [reply client/src/resources/lang/en_us.json:784]: You didn't see this comment?

File: server/src/game/end_game_condition.rs [hunk 14]
```diff
@@ -43,15 +43,12 @@ impl EndGameCondition {
...
-            let mut all_agree = true;
-            for living_role in &living_roles {
-                if !EndGameCondition::keeps_game_running(*living_role) {continue;}
-                if !EndGameCondition::required_conditions_for_win(*living_role).contains(&end_game_condition) {
-                    all_agree = false;
-                    break;
-                }
-            }
-            if all_agree {
+            //if everyone who keeps the game running agrees on this end game condition, return it
```
Comment (Jack-Papel) [inline server/src/game/end_game_condition.rs:46]: Comment is unnecessary, this is pretty obv

File: server/src/game/mod.rs [hunk 15]
```diff
@@ -272,8 +272,7 @@ impl Game {
     pub fn game_is_over(&self) -> bool {
         if let Some(_) = EndGameCondition::game_is_over(
             PlayerReference::all_players(self)
-                .filter_map(|p|
-                    if p.alive(self){Some(p.role(self))}else{None})
+                .filter_map(|p|if p.alive(self){Some(p.role(self))}else{None})
```
Comment (Jack-Papel) [inline server/src/game/mod.rs:275]: Lmao you just made it worse, whatever

File: server/src/game/role/common_role.rs (hunk 6 above)
Comment (Jack-Papel) [reply server/src/game/role/common_role.rs:138]: 'connected to end game conditions' here should probably say 'win based on their end game condition', unless this isn't the case. Either way this wording is strange

File: client/src/resources/lang/en_us.json (hunk 12 above)
Comment (Jack-Papel) [reply client/src/resources/lang/en_us.json:782]: Oh I'm dumb you addressed it

**Review by Jack-Papel — APPROVED** [review]
Address new comments

**Pull request merged**